add ems_id to container definitions and inline it in containers

moolitayer opened the pull request:
Going through joins to know what is the provider of a container is too costly

moolitayer commented:
@miq-bot add_label wip, providers/containers, enhancement
This depends on #6369

abonas commented on app/controllers/container_controller.rb:
```diff
@@ -204,7 +204,7 @@ def get_node_info(treenodeid)
     if x_node == "root" || TreeBuilder.get_model_for_prefix(@nodetype) == "MiqSearch"
       typ = "Container"
-      process_show_list
+      process_show_list(:where_clause => 'container_groups.deleted_on IS NULL')
```
how is this related to the PR's title? bad rebase? (I see 3 commits in the PR)

moolitayer replied:
This pr depends upon #6369 see [comment](https://github.com/ManageIQ/manageiq/pull/6658#issuecomment-183888470)

Fryguy commented:
Why does this change need to be built on top of #6369 ?  Tacking on an ems_id to managed entites is common in the code base, as we have it in just about every other managed entity for the exact performance reason you cited.  To me, this PR can be independent of the other.  It would be a lot easier to review as well, and is probably a no-brainer for merging.

moolitayer commented:
no longer depends on #6369

moolitayer commented:
@miq-bot remove_label wip

moolitayer commented:
@zeari @simon3z please review

zeari commented:
> Going through joins to know what is the provider of a container is too costly

Just wondering, whats the scenario for these joins? who calls for them?

zeari commented:
Also, needs tests. 
The migration wont give an ems_id to containers\definitions that are already in the db and im not sure if thats ok.

Fryguy commented:
> The migration wont give an ems_id to containers\definitions that are already in the db and im not sure if thats ok.

I assume the next refresh cleans it all up, so it's probably ok, but there might be a delay in the UI which might not be acceptable.  A data migration to fix it shouldn't be too hard.

moolitayer commented:
>> Going through joins to know what is the provider of a container is too costly
> Just wondering, whats the scenario for these joins? who calls for them?

This is mostly for reports.

moolitayer commented:
added test, changed migration script to update ems_id

simon3z commented:
@moolitayer can you check the travis failures?

moolitayer commented:
Fixed tests (mock containers now have to be created with an ems_id)
change
```belongs_to :ext_management_system, :foreign_key => "ems_id"```
to 
```belongs_to :ext_management_system, :foreign_key => :ems_id```

simon3z commented:
LGTM :+1: @chessbyte @zeari @Fryguy

moolitayer commented:
Fixed indentation in metrics_capture_spec.rb

moolitayer commented:
fixed variable name to be snake_case in migrations file

moolitayer commented:
@chessbyte @zeari @Fryguy please review

moolitayer commented:
ping @chessbyte @zeari @Fryguy please review

Fryguy commented on db/migrate/20160214115800_inline_ems_id.rb:
```diff
@@ -0,0 +1,23 @@
+    add_column :container_definitions, :ems_id, :bigint
+
+    say_with_time("Inline ems_id in containers") do
+      Container.includes(:container_group).all.each do |container|
```
Needs a stub to avoid hitting real models in migrations...if you are using associations, you will need to use a `:class_name` option as well on those.

Fryguy commented on db/migrate/20160214115800_inline_ems_id.rb:
```diff
@@ -0,0 +1,23 @@
+class InlineEmsId < ActiveRecord::Migration
```
This class needs a migration spec as it has a data migration.  When you write the migration spec you will use the `migration_stub` helper method which will make the other comment clearer because it will blow up as is.

Fryguy commented:
@moolitayer Sorry, I had not seen that you had bumped this PR.  Everything looks good except for a little bit in the migration that I commented on.  See other migration specs for examples on how to write one if you're lost.

moolitayer commented:
@Fryguy thanks for the input 

migration changes:
- use stubs
- add spec
- fixed container 'includes': 
```
+ Container.includes(:container_definitions => :container_group)
- Container.includes(:container_group)
```
- split up method to appease rubocop

@chessbyte @zeari @Fryguy please review

Fryguy commented on spec/migrations/20160214115800_inline_ems_id_spec.rb:
```diff
@@ -0,0 +1,48 @@
+  let(:container_group) { container_group_stub.create!(:ems_id => 23) }
+
+  migration_context :up do
+    it 'it adds a column called :ems_id to the containers table' do
```
No need to test functions of Rails...only the data migration part needs to be tested.

Fryguy commented on spec/migrations/20160214115800_inline_ems_id_spec.rb:
```diff
@@ -0,0 +1,48 @@
+      expect(container_stub.first.ems_id).to eq(23)
+    end
+
+    it 'it adds a column called :ems_id to the container_definitions table' do
```
This one is not needed either.

Fryguy commented on db/migrate/20160214115800_inline_ems_id.rb:
```diff
@@ -0,0 +1,51 @@
+    update_columns
+  end
+
+  def add_columns
```
I am amazed this works since the up and down are class methods, but these are instance methods.  Can you verify it works?

moolitayer replied:
It does works for some reason but I'll use instance methods for everything (I see now that the vast majority of migration use instance methods)

moolitayer commented:
address @Fryguy comments
- drop column creation check at spec
- use instance methods for up down (works but ugly)

and
- test attributes on a reloaded object created by the stub:
```
 expect(container.reload.ems_id).to eq(23)
```

@chessbyte @zeari @Fryguy please review
